### DISCUSSION TITLE
Configure multiple machines in one agent?

### DISCUSSION THREAD

**Discussion by hobbes1069** [mtconnect/cppagent#429] [Q&A]
I have several older Citizen lathes that have MTConnect but do not have a local computer to run the agent on. I would like to run `cppagent` on the same system we have Kepserver on. I see from the documentation I could create a config file per machine and install a service for each one but I would rather do it in a single service if possible/practical.

**Comment by wsobel** [comment]
You can have all the machines on the same agent. Just configure the adapter address for each machine. If you give me the info I can help with the config.

**Comment by hobbes1069** [comment]
Thanks! I'll gather up the info here soon.

**Comment by hobbes1069** [comment]
Ok, I've done some digging and the Citizen MTConnect solution is unsurprisingly using cppagent, but it's from 2018. Here's an example agent.cfg. I'd like to understand how to consolodate all the machines to one agent if that's possible.
```
Devices = C:\Program Files (x86)\DPW_L220-2 Adapter\agent\bin\Mitsubishi_M70100S174N.xml
ServiceName = DPW L220-2 Agent
Port = 5012
Adapters {
  Mitsubishi_M70100S174N {
    Host = 127.0.0.1
    Port = 7879
  }
}
Files {
    schemas {
        Path = ../schemas
        Location = /schemas/
    }
    styles {
        Path = ../styles
        Location = /styles/
    }
    Favicon {
        Path = ../styles/favicon.ico
        Location = /favicon.ico
    }
}

StreamsStyle {
  Location = /styles/Streams.xsl
}
```

**Comment by wsobel** [comment]
Add each <Device …> Section for each machine to the Device.xml

For device, list the adapter. The adapters will be on different ports and possible hosts. As follows:

```
Adapters {
  Machine1 { 
    Host = 127.0.0.1
    Port = 7879       
  }

Machine2 {
    Host = 127.0.0.1
    Port = 7878
  }
  ...
}
```

Where `Machine1` and `Machine2` are the names of the machines in the Device.xml. Each machine will need to have unique “id=…” attributes in the Device.xml. The data coming in from the adapter is keyed by name so they can be locally unique.

If you send me the Device.xml, I can send back an example.

**Comment by hobbes1069** [comment]
Ok, so I essentially want one Device.xml file with the contents of all 13 of these (the part between `<Device>...</Device>` correct?
![image](https://github.com/mtconnect/cppagent/assets/1133256/a936591e-e931-418f-9532-dd5f793da142)

Then I can add all the devices to the `Adapters` section?

I then need to configure the MQTT output.

**Comment by wsobel** [comment]
Correct. They just need unique ids in all the elements and then add the device names and their respective host/ports for the adapters. 

Tell me if you have any issues. If they are all identical or nearly identical device files, they will most likely have conflicting ids. You would need to prefix them or something to make them unique.

**Comment by hobbes1069** [comment]
I looks like they all use the machine serial number as the device ID and UUID so I should be good to go.

**Reply by wsobel** [reply]
Check each of the `id="..."` attributes. They need to be unique per w3c.

There is a way the agent can unique ids. I added the ability for dynamic device models, but it could also be run at startup. The agent will also version the device xml as well. 

Tell me if that would be helpful?

**Comment by hobbes1069** [comment]
Looks like all the attribute id's are `<machine serial num>_<incremented value>` so I think the requirement is met but I'll need to check all the xml files to be sure.

**Reply by wsobel** [reply]
That's perfect. There should be no issue.

**Reply by wsobel** [reply]
Btw. Who is the vendor for the adapter?

**Reply by hobbes1069** [reply]
I'm actually not sure. I'm based in Memphis and this is at our Mansfield facility. I'm assuming it came from the supplier of our Citizen lathes but I would need to check to be sure.

**Comment by hobbes1069** [comment]
Ok, I have one example machine setup. I noticed that I don't get the nicely formatted web page when I open it in a browser. I don't think I still fully grok how the `styles` and `schemas` work. They did have two `Streams.{xml,css}` files that I copied over but it made no difference.

In the end, it really doesn't matter since I'm converting these to MQTT machines. So how do I setup a MQTT sink for each machine?

**Reply by wsobel** [reply]
It's probably the config does not reference the style sheets. They should be in ../styles or wherever. Where are they relative to the working directory? 

You need a DevicesStyle as well. 

You can configure using path as well.

**Reply by hobbes1069** [reply]
I copied the `agent.cfg` file from the existing agent and just made minimal updates. I also copied the Streams files over from the old agent to the new. And I mis-spoke, it was `Streams.xsl` not `Streams.xml`.

I could try to sanitize the file, or perhaps I should just email you the agent.cfg?

**Reply by wsobel** [reply]
You can attach you config if you want to share or email it to me. I'll sanitize either way. Would be great to have a public example.

**Reply by hobbes1069** [reply]
Sent to the email associated on Github.

**Reply by robot-ranger** [reply]
may not be your issue, but the schema version in the stylesheet is static, and it frequently gets forgotten when rolling new mtc versions into main. When it does, it breaks the stylesheet.
We have started skipping the web agent interface, now using MetaStudio https://www.metalogi.io/products/metastudio 

(a nice piece of software and a significantly improved visual experience over the web page provided with cppagent)

**Comment by wsobel** [comment]
The latest version of the agent rewrites the stylesheet to the schema version if given. This was introduced in version 2.3. I'll check to make sure it is in the latest build.

**Reply by robot-ranger** [reply]
oh, nice good solution.
